sim/state.ts

```diff
@@ -0,0 +1,340 @@
+	// the leaf nodes like Side or Pokemon contain backreferences to Battle
+	// but don't contain the information to fill it in because the cycles in
+	// the graph have been serialized as references. Once deserialzized, the
+	// Battle can than be restarted (and provided with a `send` function for
+	// receiving updates).
+	deserializeBattle(serialized: string | /* Battle */ AnyObject): Battle {
+		const state: /* Battle */ AnyObject =
+			typeof serialized === 'string' ? JSON.parse(serialized) : serialized;
+		const options = {
+			formatid: state.format,
+			seed: state.prngSeed,
+			rated: state.rated,
+			debug: state.debugMode,
+			// We need to tell that Battle we're creating that it's been
+			// deserialized so that it allows us to populate it correctly and
+			// doesn't attempt to start playing out until we're ready.
+			deserialized: true,
+			strictChoices: state.strictChoices,
+		};
+		for (const side of state.sides) {
+			const team = side.team.split(side.team.length > 9 ? ',' : '');
+			// @ts-ignore - index signature
+			options[side.id] = {
+				name: side.name,
+				avatar: side.avatar,
+				team: team.map((p: string) => side.pokemon[Number(p) - 1].set),
+			};
+		}
+		// We create the Battle, allowing it to instantiate the Field/Side/Pokemon
+		// objects for us. The objects it creates will be incorrect, but we descend
+		// down through the fields and repopulate all of the objects with the
+		// correct state.
+		const battle = new Battle(options);
+		this.deserialize(state, battle, BATTLE, battle);
+		this.deserializeField(state.field, battle.field);
+		for (const [i, side] of state.sides.entries()) {
+			this.deserializeSide(side, battle.sides[i]);
+		}
+		battle.prng = new PRNG(state.prng);
+		// @ts-ignore - readonly
+		battle.hints = new Set(battle.hints);
+		return battle;
+	}
+
+	serializeField(field: Field): /* Field */ AnyObject {
+		return this.serialize(field, FIELD, field.battle);
+	}
+
+	private deserializeField(state: /* Field */ AnyObject, field: Field) {
+		this.deserialize(state, field, FIELD, field.battle);
+	}
+
+	serializeSide(side: Side): /* Side */ AnyObject {
+		const state: /* Side */ AnyObject = this.serialize(side, SIDE, side.battle);
+		state.pokemon = new Array(side.pokemon.length);
+		const team = new Array(side.pokemon.length);
+		for (const [i, pokemon] of side.pokemon.entries()) {
+			state.pokemon[i] = this.serializePokemon(pokemon);
+			team[side.team.indexOf(pokemon.set)] = i + 1;
+		}
+		state.team = team.join(team.length > 9 ? ',' : '');
+		state.choice = this.serializeChoice(side.choice, side.battle);
+		return state;
+	}
+
+	private deserializeSide(state: /* Side */ AnyObject, side: Side) {
+		this.deserialize(state, side, SIDE, side.battle);
+		const ordered = new Array(side.pokemon.length);
+		const team = state.team.split(state.team.length > 9 ? ',' : '');
+		for (const [i, pos] of team.entries()) {
+			ordered[Number(pos) - 1] = side.pokemon[i];
+		}
+		side.pokemon = ordered;
```

Comment:
I am not sure what you are doing here.

You seem to be restoring the Pokémon to its order at the start of the battle? Why?

Reply:
I'm doing the opposite (could use a comment). We create a `new Battle` and pass in `team`, but that creates the `pokemon` array in the wrong order. This code reshuffles the Pokemon from the starting order to the `pokemon` order.

Reply:
Oh, I see!

Still, re: my other comments, note that ``team`` doesn't match the actual ``/team`` command, after Pokémon are switched out.